File: py/okf/diff.py
"""okf — diff: a DETERMINISTIC concept-level changelog between two states of a
knowledge base.

`git diff` shows text hunks; `diff()` shows what changed as *knowledge
structure*: concepts added/removed, bodies changed (by content_hash),
frontmatter type/title changes, and graph deltas (edges added/removed, links
newly broken / fixed). Each side can be a bundle directory, a Bundle from
read_bundle(), a .duckdb catalog path, or an open duckdb connection — so
`diff(catalog, dir)` answers "what drifted since the last ingest" and
`diff(old_dir, new_dir)` compares two snapshots. Pure hash/set comparison,
outputs sorted by path: no model, no wall clock.

Mirrors r/okf/R/okf_diff.R.
"""
from __future__ import annotations
import os
from typing import Any, Optional

import duckdb

from .okf import Bundle, read_bundle, links as _links


def _state_from_con(con, bundle_id: Optional[str] = None) -> dict:
    if bundle_id is None:
        bids = [r[0] for r in con.execute(
            "SELECT DISTINCT bundle_id FROM okf_concept").fetchall()]
        if len(bids) > 1:
            raise ValueError("catalog contains multiple bundles; pass a bundle_id")
        if not bids:
            raise ValueError("catalog contains no ingested bundle")
        bundle_id = bids[0]
    cps = {p: {"type": t, "title": ti, "content_hash": h} for p, t, ti, h in con.execute(
        "SELECT path, type, title, content_hash FROM okf_concept WHERE bundle_id = ?",
        [bundle_id]).fetchall()}
    lks = con.execute(
        "SELECT src_path, dst_raw, dst_path, resolved FROM okf_link WHERE bundle_id = ?",
        [bundle_id]).fetchall()
    edges = {(s, dp) for s, _, dp, res in lks if res}
    broken = {(s, dr) for s, dr, _, res in lks if not res}
    return {"concepts": cps, "edges": edges, "broken": broken}


def bundle_state(x: Any, bundle_id: Optional[str] = None) -> dict:
    """Normalize one diff side (dir / Bundle / .duckdb path / connection) to a
    comparable state: concepts {path -> type/title/content_hash}, resolved edge
    set, and broken-link set."""
    if isinstance(x, duckdb.DuckDBPyConnection):
        return _state_from_con(x, bundle_id)
    if isinstance(x, str) and x.endswith(".duckdb") and os.path.isfile(x):
        con = duckdb.connect(x, read_only=True)
        try:
            return _state_from_con(con, bundle_id)
        finally:
            con.close()
    if isinstance(x, Bundle):
        b = x
    elif isinstance(x, str) and os.path.isdir(x):
        b = read_bundle(x)
    else:
        raise ValueError(
            "diff side must be a bundle dir, read_bundle() Bundle, .duckdb path, or connection")
    cps = {c.path: {"type": c.type, "title": c.title, "content_hash": c.content_hash}
           for c in b.concepts}
    lk = _links(b)
    edges = {(l["src_path"], l["dst_path"]) for l in lk if l["resolved"]}
    broken = {(l["src_path"], l["dst_raw"]) for l in lk if not l["resolved"]}
    return {"concepts": cps, "edges": edges, "broken": broken}


def _field_diff(common, ca, cb, field):
    out = []
    for p in sorted(common):
        x, y = ca[p][field], cb[p][field]
        if x != y:
            out.append({"path": p, "from": x, "to": y})
    return out


def _pair_diff(a: set, b: set, cols) -> list:
    return [dict(zip(cols, pair)) for pair in sorted(a - b)]


def diff(a: Any, b: Any, bundle_id_a: Optional[str] = None,
         bundle_id_b: Optional[str] = None) -> dict:
    """Concept-level diff between two states of an OKF knowledge base.

    Reports concepts added / removed / changed (by body content_hash),
    frontmatter type/title changes, and concept-graph deltas (edges added/
    removed, links newly broken or fixed). Fully deterministic: pure hash and
    set comparison, all output sorted, no wall clock.
    """
    sa = bundle_state(a, bundle_id_a)
    sb = bundle_state(b, bundle_id_b)
    ca, cb = sa["concepts"], sb["concepts"]

    added = sorted(set(cb) - set(ca))
    removed = sorted(set(ca) - set(cb))
    common = set(ca) & set(cb)
    changed = sorted(p for p in common if ca[p]["content_hash"] != cb[p]["content_hash"])

    type_changed = _field_diff(common, ca, cb, "type")
    retitled = _field_diff(common, ca, cb, "title")

    links_added = _pair_diff(sb["edges"], sa["edges"], ("src_path", "dst_path"))
    links_removed = _pair_diff(sa["edges"], sb["edges"], ("src_path", "dst_path"))
    broken_added = _pair_diff(sb["broken"], sa["broken"], ("src_path", "dst_raw"))
    broken_fixed = _pair_diff(sa["broken"], sb["broken"], ("src_path", "dst_raw"))

    identical = not (added or removed or changed or type_changed or retitled or
                     links_added or links_removed or broken_added or broken_fixed)

    return {
        "identical": identical, "added": added, "removed": removed, "changed": changed,
        "type_changed": type_changed, "retitled": retitled,
        "links_added": links_added, "links_removed": links_removed,
        "broken_added": broken_added, "broken_fixed": broken_fixed,
        "summary": {
            "added": len(added), "removed": len(removed), "changed": len(changed),
            "unchanged": len(common) - len(changed),
            "type_changed": len(type_changed), "retitled": len(retitled),
            "links_added": len(links_added), "links_removed": len(links_removed),
            "broken_added": len(broken_added), "broken_fixed": len(broken_fixed)},
    }

File: cpp/okf/diff.cpp
// Deterministic concept-level diff — mirrors py/okf/diff.py. Pure hash/set
// comparison, all output sorted. Drift mode: diff(DiffSide::from_ingested(i),
// DiffSide::from_dir(d)) answers "what drifted since the ingest" catalog-free.

#include <map>

#include "okf.hpp"

namespace okf {
namespace {

struct ConceptMeta {
    std::optional<std::string> type, title;
    std::string content_hash;
};

struct State {
    std::map<std::string, ConceptMeta> concepts;
    std::set<std::pair<std::string, std::string>> edges, broken;
};

State state_from(const Bundle& b, const std::vector<Link>& lk) {
    State s;
    for (const Concept& c : b.concepts) {
        s.concepts[c.path] = ConceptMeta{c.type, c.title, c.content_hash};
    }
    for (const Link& l : lk) {
        if (l.resolved) s.edges.insert({l.src_path, l.dst_path.value_or("")});
        else s.broken.insert({l.src_path, l.dst_raw});
    }
    return s;
}

State bundle_state(const DiffSide& x) {
    if (x.ingested) return state_from(x.ingested->bundle, x.ingested->links);
    if (x.bundle) return state_from(*x.bundle, links(*x.bundle));
    if (x.dir) {
        Bundle b = read_bundle(*x.dir, "dir");
        return state_from(b, links(b));
    }
    throw OkfError("diff side must be a bundle dir, Bundle, or Ingested");
}

std::vector<FieldChange> field_diff(const State& sa, const State& sb,
                                    std::optional<std::string> ConceptMeta::* field) {
    std::vector<FieldChange> out;
    for (const auto& [path, ma] : sa.concepts) {
        auto it = sb.concepts.find(path);
        if (it == sb.concepts.end()) continue;
        const auto& va = ma.*field;
        const auto& vb = it->second.*field;
        if (va != vb) out.push_back(FieldChange{path, va, vb});
    }
    return out;
}

std::vector<EdgeDelta> pair_diff(const std::set<std::pair<std::string, std::string>>& a,
                                 const std::set<std::pair<std::string, std::string>>& b) {
    std::vector<EdgeDelta> out;
    for (const auto& pr : a) {
        if (!b.count(pr)) out.push_back(EdgeDelta{pr.first, pr.second});
    }
    return out;
}

}  // namespace

Diff diff(const DiffSide& a, const DiffSide& b) {
    State sa = bundle_state(a);
    State sb = bundle_state(b);

    Diff d;
    for (const auto& [path, meta] : sb.concepts) {
        if (!sa.concepts.count(path)) d.added.push_back(path);
    }
    for (const auto& [path, meta] : sa.concepts) {
        auto it = sb.concepts.find(path);
        if (it == sb.concepts.end()) d.removed.push_back(path);
        else if (meta.content_hash != it->second.content_hash) d.changed.push_back(path);
    }
    d.type_changed = field_diff(sa, sb, &ConceptMeta::type);
    d.retitled = field_diff(sa, sb, &ConceptMeta::title);
    d.links_added = pair_diff(sb.edges, sa.edges);
    d.links_removed = pair_diff(sa.edges, sb.edges);
    d.broken_added = pair_diff(sb.broken, sa.broken);
    d.broken_fixed = pair_diff(sa.broken, sb.broken);
    d.identical = d.added.empty() && d.removed.empty() && d.changed.empty() &&
                  d.type_changed.empty() && d.retitled.empty() && d.links_added.empty() &&
                  d.links_removed.empty() && d.broken_added.empty() && d.broken_fixed.empty();
    return d;
}

}  // namespace okf
